### PR TITLE
Add support for CPU sockets and threads

### PR DESCRIPTION
**Pull request by ksimon1**
**What this PR does / why we need it**:
Add support for CPU sockets and threads

https://trello.com/c/W8zaxGZp/93-8epic-perf-baselinebug-add-support-for-cpu-sockets-and-threads-initiative-perfscale-baseline

@MarSik, @fromanirh, @yanirq, @petrkotas, @rmohr please review

**Release note**:
```release-note
Add support for CPU sockets and threads
```

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically.
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

File: pkg/api/v1/openapi_generated.go [hunk 1]
```diff
@@ -156,6 +156,20 @@ func schema_kubevirt_pkg_api_v1_CPU(ref common.ReferenceCallback) common.OpenAPI
 							Format:      "int64",
 						},
 					},
+					"sockets": {
+						SchemaProps: spec.SchemaProps{
+							Description: "Sockets specifies the number of sockets inside the vmi. Must be a value greater or equal 1.",
+							Type:        []string{"integer"},
+							Format:      "int64",
```
Comment (fromanirh) [inline pkg/api/v1/openapi_generated.go]: I wonder if we should use unsigned integer. Not sure what the kubevirt best practices are.

File: pkg/api/v1/openapi_generated.go [hunk 2]
```diff
@@ -156,6 +156,20 @@ func schema_kubevirt_pkg_api_v1_CPU(ref common.ReferenceCallback) common.OpenAPI
...
+						},
+					},
+					"threads": {
+						SchemaProps: spec.SchemaProps{
+							Description: "Threads specifies the number of threads inside the vmi. Must be a value greater or equal 1.",
+							Type:        []string{"integer"},
+							Format:      "int64",
```
Comment (fromanirh) [inline pkg/api/v1/openapi_generated.go]: same

File: pkg/api/v1/schema.go [hunk 3]
```diff
@@ -158,6 +158,12 @@ type CPU struct {
 	// Cores specifies the number of cores inside the vmi.
 	// Must be a value greater or equal 1.
 	Cores uint32 `json:"cores,omitempty"`
+	// Sockets specifies the number of sockets inside the vmi.
+	// Must be a value greater or equal 1.
+	Sockets uint32 `json:"sockets,omitempty"`
```
Comment (fromanirh) [inline pkg/api/v1/schema.go]: ...especially considering that you are using (correctly, seeing the surrounding code) unsigned integers here.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 4]
```diff
@@ -1018,6 +1018,22 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
...
+}
+
+func calculateRequestedThreads(vmi *v1.VirtualMachineInstance) uint32 {
+	threads := uint32(1)
+	if vmi.Spec.Domain.CPU != nil && vmi.Spec.Domain.CPU.Threads != 0 {
+		return vmi.Spec.Domain.CPU.Threads
+	}
+	return threads
+}
+
```
Comment (fromanirh) [inline pkg/virt-launcher/virtwrap/api/converter.go]: I think we need to make sure that sockets, threads, cores (and any combination thereof) falls under the constraints set by the pod. AFAIK/R, kubernetes only knows about (fractions of) cores, considering them like execution unit.
Here we are shuffling things around, meaning that, given for example 2 kubevirt "cores" (execution unit) we can either consider them
- 2 thread 1 core 1 socket CPU
- 1 thread 2 core 1 socket CPU
- 1 thread 1 core 2 socket CPU

things can get messy pretty fast (what about 16 kubevirt "cores" :)). I think we should at least laid foundations to ensure we adhere to the limits (eg. we can't somehow get to have 16 threads 16 cores 16 socket CPUs)

**Comment by yanirq** [comment]
- Are we going to perform validations on CPU threads/sockets ?   e.g. number of threads against cores, combination of values for cores/threads/sockets and how the system performs after its converter in domxml  and launched ?   (doesn't necessarily have to be under this PR) 

- make sure you run make generate before committing and pushing all the related files.  just to see everything is in place (api change)

**Pull request closed without merging**